connmgr.go

```diff
@@ -43,6 +43,27 @@ type BasicConnMgr struct {
 
 var _ ifconnmgr.ConnManager = (*BasicConnMgr)(nil)
 
+type segment struct {
+	sync.Mutex
+	peers map[peer.ID]*peerInfo
+}
+
+type segments [256]*segment
+
+func (s *segments) get(id peer.ID) *segment {
+	b := []byte(id)
+	return s[b[len(b)-1]]
+}
+
+func (s *segments) countPeers() (count int) {
+	for _, seg := range s {
+		seg.Lock()
```

Comment:
This can be a reader lock.

Reply:
there isn't a large number of readers on this though.